### PR TITLE
Ohm 1052 reduce docker image size

### PR DESCRIPTION
**Pull request by Ken473**

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -1 +1,4 @@
-node_modules
\ No newline at end of file
+node_modules
+infrastructure
+packaging
+test
```
Comment (MattyJ007) [inline .dockerignore:4]: Please add a newline to the end of the file

File: Dockerfile [hunk 2]
```diff
@@ -1,23 +1,18 @@
 # Openhim Console Dockerfile for latest changes
-FROM ubuntu:16.04
+FROM node:14.17-alpine
```
Comment (MattyJ007) [inline Dockerfile:2]: This should be nginx:alpine

File: Dockerfile [hunk 3]
```diff
@@ -1,23 +1,18 @@
...
+RUN apk upgrade --update-cache --available
+RUN apk add git
```
Comment (MattyJ007) [inline Dockerfile:4]: Why is this here?

File: Dockerfile [hunk 4]
```diff
@@ -1,23 +1,18 @@
...
-RUN git clone https://github.com/jembi/openhim-console.git
-
-WORKDIR /etc/openhim-console
+COPY package.json /app
+COPY . .
 
 # Install dependencies and build
-RUN npm i
+RUN npm install
+RUN npm install webpack-dev-server && npm install webpack webpack-cli --save-dev
```
Comment (MattyJ007) [inline Dockerfile:14]: These are dev dependencies, are they needed for the prod build?

File: infrastructure/docker-compose.yml [hunk 5]
```diff
@@ -14,7 +14,7 @@ services:
 
   openhim-core:
     container_name: openhim-core
-    image: jembi/openhim-core:latest
+    image: openhim-alpinev3-multi-staged
```
Comment (MattyJ007) [inline infrastructure/docker-compose.yml:17]: This is not a valid image. Please revert

File: infrastructure/docker-compose.yml [hunk 6]
```diff
@@ -34,7 +34,7 @@ services:
 
   openhim-console:
     container_name: openhim-console
-    image: jembi/openhim-console:latest
+    image: openhim-console-alpinev1
```
Comment (MattyJ007) [inline infrastructure/docker-compose.yml:37]: This is not a valid image. Please revert

File: Dockerfile [hunk 7]
```diff
@@ -1,23 +1,18 @@
...
-RUN apt-get install -y bzip2
-RUN apt-get install -y curl
-RUN curl -sL https://deb.nodesource.com/setup_8.x | bash -
-RUN apt-get install -y nodejs
-
-# Clone Openhim-console repo
-RUN git clone https://github.com/jembi/openhim-console.git
-
-WORKDIR /etc/openhim-console
+COPY package.json /app
```
Comment (MattyJ007) [inline Dockerfile:9]: Nginx will serve html files in this dir `/usr/share/nginx/html`

**Review by MattyJ007 — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 3 above)
Comment (Ken473) [reply Dockerfile:4]: I found some dependencies in the package.json file that referenced git repositories. Without these the application would not build.

File: Dockerfile (hunk 2 above)
Comment (Ken473) [reply Dockerfile:2]: duly noted. Will change these

File: Dockerfile (hunk 3 above)
Comment (MattyJ007) [reply Dockerfile:4]: I see, so then I guess perhaps the node alpine can be part of the multi-stage build. Then just copy over the `/dist` directory into the alpine image

File: Dockerfile (hunk 4 above)
Comment (MattyJ007) [reply Dockerfile:14]: I see they are needed but the regular install will include them

File: Dockerfile [hunk 8]
```diff
@@ -1,23 +1,18 @@
...
+COPY . .
 
 # Install dependencies and build
-RUN npm i
+RUN npm install
+RUN npm install webpack-dev-server && npm install webpack webpack-cli --save-dev
+
```
Comment (MattyJ007) [inline Dockerfile:15]: The two lines below can be removed since `npm install` covers *npm build* (it automatically triggers the `npm run prepare` script).
 and nginx will serve the HTML without the http-server

**Review by MattyJ007 — CHANGES_REQUESTED** [review]

File: infrastructure/docker-compose.yml (hunk 6 above)
Comment (Ken473) [reply infrastructure/docker-compose.yml:37]: sure

File: infrastructure/docker-compose.yml (hunk 5 above)
Comment (Ken473) [reply infrastructure/docker-compose.yml:17]: sure

File: Dockerfile (hunk 7 above)
Comment (MattyJ007) [reply Dockerfile:9]: Yes, just run `npm install` and that will take care of everything

File: Dockerfile (hunk 4 above)
Comment (Ken473) [reply Dockerfile:14]: We should change the npm run build to npm run build prod. This one includes the dev dependencies and I do not know if it includes anything that may affect you in production

File: Dockerfile (hunk 7 above)
Comment (Ken473) [reply Dockerfile:9]: cool

File: infrastructure/docker-compose.yml (hunk 5 above)
Comment (MattyJ007) [reply infrastructure/docker-compose.yml:17]: This has not been reverted

File: infrastructure/docker-compose.yml (hunk 6 above)
Comment (MattyJ007) [reply infrastructure/docker-compose.yml:37]: This has not been reverted

File: Dockerfile [hunk 9]
```diff
@@ -1,25 +1,18 @@
...
-RUN npm i -g http-server 
+RUN npm install
+
+FROM nginx:mainline-alpine
+WORKDIR /usr/share/nginx/html
 
-# Host and run server
-CMD http-server ./dist -p 80
\ No newline at end of file
+COPY --from=build /app/dist  /usr/share/nginx/html
```
Comment (MattyJ007) [inline Dockerfile:18]: Add new line to end of file

File: Dockerfile (hunk 7 above)
Comment (MattyJ007) [reply Dockerfile:9]: Why is this step separate from `COPY . .` ?

File: Dockerfile [hunk 10]
```diff
@@ -1,25 +1,18 @@
 # Openhim Console Dockerfile for latest changes
-FROM ubuntu:16.04
+FROM node:14.17-alpine as build
+RUN apk upgrade --update-cache --available
```
Comment (MattyJ007) [inline Dockerfile:3]: Is this step necessary?

**Review by MattyJ007 — CHANGES_REQUESTED** [review]
Looks good - just a few minor comments

File: Dockerfile (hunk 3 above)
Comment (Ken473) [reply Dockerfile:4]: That is done

File: Dockerfile (hunk 10 above)
Comment (Ken473) [reply Dockerfile:3]: I had added it to make sure we are not missing any required dependencies. Let me test and see what's happening

File: Dockerfile (hunk 7 above)
Comment (Ken473) [reply Dockerfile]: forgot to remove it. Should be gone now.

File: Dockerfile (hunk 10 above)
Comment (Ken473) [reply Dockerfile]: Looks good so far.

File: infrastructure/docker-compose.yml (hunk 5 above)
Comment (Ken473) [reply infrastructure/docker-compose.yml]: Please check new commit

File: infrastructure/docker-compose.yml (hunk 6 above)
Comment (Ken473) [reply infrastructure/docker-compose.yml]: same as above

**Review by MattyJ007 — APPROVED** [review]
